config/html-webpack-inline-critical-css-plugin.js

```diff
@@ -0,0 +1,255 @@
+          });
+      });
+    });
+  }
+
+  processStyle(style, document) {
+    let done = Promise.resolve();
+    let sheet = style.childNodes.length>0 && style.childNodes.map(getNodeValue).join('\n');
+    if (!sheet) return done;
+
+    const ast = css.parse(sheet);
+
+    const before = sheet;
+
+    visit(ast, function (rule) {
+      if (rule.type==='rule') {
+        rule.selectors = rule.selectors.filter(function (sel) {
+          if (sel.match(/:(hover|focus|active)([.[#~&^:*]|\s|\n|$)/)) return false;
+          sel = sel.replace(/::?(?:[a-z-]+)([.[#~&^:*]|\s|\n|$)/gi, '$1');
+          return document.querySelector(sel, document) != null;
+        });
+        // If there are no matched selectors, replace the rule with a comment.
+        if (rule.selectors.length===0) {
+          rule.type = 'comment';
+          rule.comment = '';
+          delete rule.selectors;
+          delete rule.declarations;
+        }
+      }
+
+      if (rule.rules) {
+        // Filter out comments
+        rule.rules = rule.rules.filter(notComment);
+        // If there are no remaining rules, replace the parent rule with a comment.
+        if (rule.rules.length===0) {
+          rule.type = 'comment';
+          rule.comment = '';
+          delete rule.rules;
+        }
+      }
+    });
+
+    sheet = css.stringify(ast, { compress: true });
+
+    if (this.options.minimize || this.options.compress || this.options.minify) {
+      const cssnano = require('cssnano');
+      done = cssnano.process(sheet, {}, { preset: 'default' }).then(function (result) {
+        sheet = result.css;
+      });
+    }
+
+    return done.then(function () {
+      if (sheet.trim().length===0) {
+        // all rules were removed, get rid of the style element entirely
+        sheet.parentNode.removeChild(sheet);
+      }
+      else {
+        // replace the stylesheet inline
+        while (style.lastChild) {
+          style.removeChild(style.lastChild);
+        }
+        style.appendChild(document.createTextNode(sheet));
+      }
+      const name = style.$$name;
+      console.log('\u001b[32mInlined CSS Size' + (name ? (' ('+name+')') : '') + ': ' + (sheet.length / 1000).toPrecision(2) + 'kb (' + ((before.length - sheet.length) / before.length * 100 | 0) + '% of original ' + (before.length / 1000).toPrecision(2) + 'kb)\u001b[39m');
+    });
+  }
+};
+
+
+function visit(node, fn) {
+  if (node.stylesheet) return visit(node.stylesheet, fn);
+
+  node.rules.forEach(function (rule) {
+    if (rule.rules) {
+      visit(rule, fn);
+    }
+    fn(rule);
+  });
+}
+
+
+function readFile(file) {
+  return new Promise(function (resolve, reject) {
+    fs.readFile(file, 'utf8', function (err, contents) {
+      if (err) reject(err);
+      else resolve(contents);
+    });
+  });
+}
+
+function getNodeValue(node) {
+  return node.nodeValue;
+}
+
+function notComment(rule) {
+  return rule.type !== 'comment';
+}
+
+function getElementsByTagName(tagName) {
```

Comment:
I mean, it doesn’t really matter. But a recursive function could be easier to read:

```js
function getElementsByTagName(tagName) {
  if (this.nodeType !== 1) return []; // Only elements are relevant
  tagName = tagName.toUpperCase();
  return [
    ...((this.tagName == tagName || tagName === '*') ? tagName : [])), // Add current element if it matches tag
    ...this.children.map(child => getElementByTagName(tagName).call(child)) // Check children recursively
  ];
}
```

Reply:
easier to read but quite a bit slower. also the weird edge cases are to handle the fact that this dom doesn't normalize the case of tagName

Reply:
Whoop, here I was spouting "facts" without testing. Recursion is *much* faster here:

<a href="https://esbench.com/bench/5ac3b647f2949800a0f619e1"><img width="502" src="https://user-images.githubusercontent.com/105127/38267550-e2b42a56-3749-11e8-9530-e139bb62dc42.png"></a>

I'll update with the ones from that [bench](https://esbench.com/bench/5ac3b647f2949800a0f619e1).

Reply:
Even if it wasn’t, I’d prefer the more readable/concise version (unless the difference is on the magnitude of seconds). It’s a build step after all.

Reply:
Also we can probably safe the `bind()` and turn the element into a normal parameter, if that makes it more readable.

Reply:
The bind is because this gets mixed into Element.prototype.

My benchmark was flawed, stack is around 5x faster, or 2x faster than an optimized recursive version:

<img width="402" src="https://user-images.githubusercontent.com/105127/38273279-95eb97c6-3759-11e8-9d36-7550c8aab085.png">

(same url: https://esbench.com/bench/5ac3b647f2949800a0f619e1)